api/prometheus/v1/api.go

```diff
@@ -169,6 +253,56 @@ type queryResult struct {
 	v model.Value
 }
 
+func (rg *RuleGroup) UnmarshalJSON(b []byte) error {
+	v := struct {
+		Name     string                   `json:"name"`
+		File     string                   `json:"file"`
+		Interval float64                  `json:"interval"`
+		Rules    []map[string]interface{} `json:"rules"`
+	}{}
+
+	if err := json.Unmarshal(b, &v); err != nil {
+		return err
+	}
+
+	rg.Name = v.Name
+	rg.File = v.File
+	rg.Interval = v.Interval
+
+	for _, rule := range v.Rules {
+		// Because both recording and alerting rules are stored in the same
+		// JSON array, each rule is first encoded into JSON and then decoded
+		// into either a RecordingRule or AlertingRule.
+		t, ok := rule["type"]
+		if !ok {
+			return errors.New("rule has no type field")
+		}
+		ruleJSON, err := json.Marshal(rule)
```

Comment:
I am a bit confused here. You are marshaling a rule to unmarshal it?

How about unmarshaling the rule into something like this first:

```golang
struct {
  type string
}
```

and depending on `type` unmarshal as a recording or an alerting rule?

Reply:
With that approach I think we'd still need to get the complete byte or JSON encoding of each individual rule (by using MarshalJSON) in order to unmarshal them into a typed recording or alerting rule. That said, I think most of the complexity here is from decoding the mixed type slice and there may not be a super elegant way to do it in Go.

Reply:
I've made an update to this PR that should be easier to read and is more in-line with the approach suggested in this comment.